program_transformers/src/rollups/rollup_persister.rs

```diff
@@ -0,0 +1,571 @@
+    async fn download_rollup_and_check_checksum(
+        &self,
+        url: &str,
+        checksum: &str,
+    ) -> Result<Box<Rollup>, RollupValidationError>;
+}
+
+pub struct RollupPersister<T: ConnectionTrait + TransactionTrait, D: RollupDownloader> {
+    txn: Arc<T>,
+    downloader: D,
+    cl_audits: bool,
+}
+
+pub struct RollupDownloaderForPersister {}
+
+#[async_trait]
+impl RollupDownloader for RollupDownloaderForPersister {
+    async fn download_rollup(&self, url: &str) -> Result<Box<Rollup>, RollupValidationError> {
+        let response = reqwest::get(url).await?.bytes().await?;
+        Ok(Box::new(serde_json::from_slice(&response)?))
+    }
+
+    async fn download_rollup_and_check_checksum(
+        &self,
+        url: &str,
+        checksum: &str,
+    ) -> Result<Box<Rollup>, RollupValidationError> {
+        let response = reqwest::get(url).await?.bytes().await?;
+        let file_hash = xxhash_rust::xxh3::xxh3_128(&response);
+        let hash_hex = hex::encode(file_hash.to_be_bytes());
+        if hash_hex != checksum {
+            return Err(RollupValidationError::InvalidDataHash(
+                checksum.to_string(),
+                hash_hex,
+            ));
+        }
+        Ok(Box::new(serde_json::from_slice(&response)?))
+    }
+}
+
+impl<T: ConnectionTrait + TransactionTrait, D: RollupDownloader> RollupPersister<T, D> {
+    pub fn new(txn: Arc<T>, downloader: D, cl_audits: bool) -> Self {
+        Self {
+            txn,
+            downloader,
+            cl_audits,
+        }
+    }
+
+    pub async fn persist_rollups(&self) {
+        loop {
+            tokio::time::sleep(Duration::from_secs(5)).await;
+            let Ok((rollup_to_verify, rollup)) = self.get_rollup_to_verify().await else {
+                continue;
+            };
+            let Some(rollup_to_verify) = rollup_to_verify else {
+                // no rollups to persist
+                continue;
+            };
+            let Ok(rollup) = rollup
+                .map(|r| bincode::deserialize::<Rollup>(r.rollup_binary_bincode.as_slice()))
+                .transpose()
+                .map_err(|e| ProgramTransformerError::DeserializationError(e.to_string()))
+            else {
+                continue;
+            };
+            self.persist_rollup(rollup_to_verify, rollup.map(Box::new))
+                .await;
+        }
+    }
+
+    pub async fn persist_rollup(
+        &self,
+        mut rollup_to_verify: rollup_to_verify::Model,
+        mut rollup: Option<Box<Rollup>>,
+    ) {
+        info!("Persisting {} rollup", &rollup_to_verify.url);
+        loop {
+            match &rollup_to_verify.rollup_persisting_state {
+                &RollupPersistingState::ReceivedTransaction => {
+                    if let Err(err) = self
+                        .download_rollup(&mut rollup_to_verify, &mut rollup)
+                        .await
```

Comment:
I think with most DAS processing there is an expectation that there will eventually be multiple concurrent processors of the data in order to handle volume.  So in this case, there would be multiple consumers of the same `rollup_to_verify` table row, and it seems like we would need to lock an entry or have some other means to prevent a race condition of multiple processors both downloading and then changing to the next state.  The background task processing has a lock state, which may not be perfect.  I may be misunderstanding but it seems like this processing also needs a plan for locking.

Reply:
Thank you for noticing this! I added  RollupPersistingState::StartProcessing. So now when i select rollup to process i set state for this rollup to StartProcessing inside single transaction. i hope this will prevent such case as you described, wdyt?

Reply:
I think yes `StartProcessing` state will be able to prevent issues with multiple processors in `persist_rollup()`, but it must be slightly changed to not do any handling of items in `ReceivedTransaction` state.  See other comment for more information where I commented on `get_rollup_to_verify()` based on my study and understanding of the code.